pkg/controllermanager/controller/shoot/shoot_control_delete.go

```diff
@@ -481,48 +460,28 @@ func (c *defaultControl) updateShootStatusDeleteError(o *operation.Operation, la
+func (c *defaultControl) needsControlPlaneDeployment(o *operation.Operation) (bool, error) {
+	// Get the infrastructure resource
+	infrastructure := &extensionsv1alpha1.Infrastructure{}
+	if err := o.K8sSeedClient.Client().Get(context.TODO(), kutil.Key(o.Shoot.SeedNamespace, o.Shoot.Info.Name), infrastructure); err != nil {
 		if apierrors.IsNotFound(err) {
-			// The infrastructure resource has not been found - we need to check whether the Terraform state does still exist.
-			// If it does still exist then we need to migrate. Otherwise there are no infrastructures resources anymore that
-			// need to be deleted, so no migration would be necessary.
-
-			tf, err := o.NewShootTerraformer(common.TerraformerPurposeInfraDeprecated)
-			if err != nil {
-				return false, err
-			}
-
-			configExists, err := tf.ConfigExists()
-			if err != nil {
-				return false, err
-			}
-
-			return configExists, nil
+			// The infrastructure resource has not been found, no need to redeploy the control plane
+			return false, nil
 		}
 		return false, err
 	}
 
-	// The infrastructure resource has been found - no need to migrate as it already exists.
-	return false, nil
-}
+	// The infrastructure resource has been found, so we should redeploy the control plane depending on its status.
+	if lastOperation := infrastructure.Status.LastOperation; lastOperation != nil && lastOperation.State == gardencorev1alpha1.LastOperationStateSucceeded {
 
-func (c *defaultControl) needsWorkerMigration(o *operation.Operation) (bool, error) {
-	if err := o.K8sSeedClient.Client().Get(context.TODO(), kutil.Key(o.Shoot.SeedNamespace, o.Shoot.Info.Name), &extensionsv1alpha1.Worker{}); err != nil {
-		if apierrors.IsNotFound(err) {
-			// The Worker resource has not been found - we need to check whether the MCM deployment does still exist.
-			// If it does still exist then we need to migrate. Otherwise there are no machine resources anymore that
-			// need to be deleted, so no migration would be necessary.
-			machineDeployments := &machinev1alpha1.MachineDeploymentList{}
-			if err := o.K8sSeedClient.Client().List(context.TODO(), &client.ListOptions{Raw: &metav1.ListOptions{Limit: 1}}, &machinev1alpha1.MachineDeploymentList{}); err != nil {
-				return false, err
-			}
-			return len(machineDeployments.Items) != 0, nil
+		// Get the infrastructure status from the infrastructure resource
+		if providerStatus := infrastructure.Status.ProviderStatus; providerStatus != nil {
+			o.Shoot.InfrastructureStatus = providerStatus.Raw
+			return true, nil
 		}
-
-		return false, err
 	}
 
-	// The Worker resource has been found - no need to migrate as it already exists.
+	// The infrastructure resource has been found, but its status is not Succeeded or its provider status is nil
+	// If the infrastructure provisioning failed, the control plane was not created at all, so no need to redeploy it
```

Comment:
it might also be the case that it is currently reconciled (processing).
I'd suggest to only check if the provider status is non-nil and if so then to return `true`. If the provider status is non-nil it means that the infrastructure creation succeeded at least once and we can just take the generated status. We probably don't need to care about the last operation status.

Reply:
Done